backend/backend.go

```diff
@@ -122,3 +122,38 @@ func UnmarshalHistoryEvent(bytes []byte) (*HistoryEvent, error) {
 	}
 	return e, nil
 }
+
+func GetSubOrchestrationInstances(ctx context.Context, be Backend, iid api.InstanceID, isPurge bool) ([]api.InstanceID, error) {
+	owi := &OrchestrationWorkItem{
+		InstanceID: iid,
+	}
+	state, err := be.GetOrchestrationRuntimeState(ctx, owi)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch orchestration state for '%s': %w", iid, err)
+	}
+	if isPurge {
+		if len(state.NewEvents())+len(state.oldEvents) == 0 {
+			// If there are no events, the orchestration instance doesn't exist
+			return nil, api.ErrInstanceNotFound
+		}
+		if !state.IsCompleted() {
+			// Orchestration must be completed before purging its state
+			return nil, api.ErrNotCompleted
+		}
+	}
+	oldEvents := state.OldEvents()
+	newEvents := state.NewEvents()
+
+	subOrchestrationInstances := make([]api.InstanceID, 0, 10)
+	for _, e := range oldEvents {
+		if created := e.GetSubOrchestrationInstanceCreated(); created != nil {
+			subOrchestrationInstances = append(subOrchestrationInstances, api.InstanceID(created.InstanceId))
+		}
+	}
+	for _, e := range newEvents {
+		if created := e.GetSubOrchestrationInstanceCreated(); created != nil {
+			subOrchestrationInstances = append(subOrchestrationInstances, api.InstanceID(created.InstanceId))
+		}
+	}
+	return subOrchestrationInstances, nil
```

Comment:
Just like in my previous implementation, we need to filter out all instances that have already completed. See the logic I used [here](https://github.com/microsoft/durabletask-go/blob/aa335e20a192395649eb9ede97b57c9f647238ea/task/orchestrator.go#L546-L565).

Reply:
Actually that's intentional. We should send terminate event to even completed/failed instances, as there might be a case that an orchestration has completed/failed, but it's subOrchestration might still be running and so we would need to terminate those as well. See the second test added in orchestration_test.go